feat(cli): Views Introspection Warnings

Druue opened the pull request:
Added new warnings for views introspection:
- 21: IntrospectionWarningsUnsupportedTypesInViews
- 22: IntrospectionWarningEnrichedWithMapOnFieldInViews
- 23: IntrospectionWarningEnrichedWithMapOnView
- 24: IntrospectionWarningViewsWithoutIdentifier
- 25: IntrospectionWarningEnrichedWithCustomPrimaryKeyNamesInViews
- 26: IntrospectionWarningsFieldsWithEmptyNamesInViews

closes #17578

Commit:
added warnings types

Commit:
added types to union

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/prisma/prisma?pullRequest=18000) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/prisma/prisma?pullRequest=18000) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/prisma/prisma?pullRequest=18000) it.</sub>

Commit:
views warning handling

Druue commented:
hm, we seem to lose type information here when using includes. but also I don't see an easy way to just add the annotation manually. Do we want to update the imports to add that annotation? or should I manually check equality on all of the codes in this case
<img width="787" alt="image" src="https://user-images.githubusercontent.com/29753584/220105069-67ed5228-afce-492a-ab76-692d71366a2c.png">

Commit:
added missing warning code 26

Druue commented:
## Warning Code 21
```sql
create table reservations(
  id int primary key,
  room varchar not null,
  dates daterange not null 
);

create view res as select id, room, dates from reservations;
```

add to your `schema.prisma` -- otherwise the following warning message will be obfuscated by a warning that the view has no PK
```prisma
view res {
  id    Int @id
}
```

then running db pull will yield:
```prisma
view res {
  id    Int                       @id
  room  String?                   @db.VarChar
  dates Unsupported("daterange")?
}
```

and the following output in the cli
<img width="786" alt="image" src="https://user-images.githubusercontent.com/29753584/220342933-996c02c9-d8c0-43ab-a4b1-963ce375675b.png">

Druue commented:
## Warning Code 22
```sql
CREATE VIEW "A" AS SELECT 1 AS id;
```

running db pull will yield:
```prisma
/// The underlying view does not contain a valid unique identifier and can therefore currently not be handled by the Prisma Client.
view A {
  id Int?

  @@ignore
}
```

and the following output in the cli
<img width="640" alt="image" src="https://user-images.githubusercontent.com/29753584/220570962-efdc5d98-083a-482a-b70a-3790b75251a3.png">

Druue commented:
## Warning Code 23
```sql
CREATE VIEW "if" AS SELECT 1 AS id;
```

running db pull will yield:
```prisma
/// This view has been renamed to 'Renamedif' during introspection, because the original name 'if' is reserved.
/// The underlying view does not contain a valid unique identifier and can therefore currently not be handled by the Prisma Client.
view Renamedif {
  id Int @id

  @@map("if")
  @@ignore
}
```

and the following output in the cli -- note this cli output is obfuscated by the lack of id attribute so one would first need to manually add the following to their schema to get this output.
```prisma
view Renamedif {
  id Int @id

  @@map("if")
}
```
<img width="639" alt="image" src="https://user-images.githubusercontent.com/29753584/220573231-c9ded487-d9c3-4f89-a406-7945b5b8b53b.png">

Druue commented:
## Warning Code 24
```sql
CREATE TABLE "User" (
  id SERIAL PRIMARY KEY,
  first_name VARCHAR(255) NOT NULL,
  last_name VARCHAR(255) NULL
);

CREATE VIEW "Schwuser" AS
  SELECT id, first_name, last_name FROM "User";
```

running db pull will yield:
```prisma
model User {
  id         Int     @id @default(autoincrement())
  first_name String  @db.VarChar(255)
  last_name  String? @db.VarChar(255)
}

/// The underlying view does not contain a valid unique identifier and can therefore currently not be handled by the Prisma Client.
view Schwuser {
  id         Int?
  first_name String? @db.VarChar(255)
  last_name  String? @db.VarChar(255)

  @@ignore
}
```

and the following output in the cli
<img width="765" alt="image" src="https://user-images.githubusercontent.com/29753584/220575532-26259893-3b64-4f73-9a7c-c00e51cc5308.png">

Druue commented:
## Warning Code 25
```sql
CREATE VIEW "B" AS SELECT 1 AS a, 2 AS b;
```

running db pull will yield:
```prisma
/// The underlying view does not contain a valid unique identifier and can therefore currently not be handled by the Prisma Client.
view B {
  a Int?
  b Int?

  @@ignore
}
```

and the following output in the cli -- note this cli output is obfuscated by the lack of id attribute so one would first need to manually add the following to their schema to get this output.
```prisma
view B {
  a Int
  b Int

  @@id([a, b], name: "kekw")
}
```
<img width="673" alt="image" src="https://user-images.githubusercontent.com/29753584/220576757-57dd9d3f-446b-419a-a85f-f8987f9ca147.png">

Druue commented:
## Warning Code 26
```sql
CREATE VIEW "A" AS SELECT 2 AS foo, 1 AS "1";
```

running db pull will yield:
```prisma
/// The underlying view does not contain a valid unique identifier and can therefore currently not be handled by the Prisma Client.
view C {
  foo Int?

  /// This field was commented out because of an invalid name. Please provide a valid one that matches [a-zA-Z][a-zA-Z0-9_]*
  // 1 Int? @map("1")
  @@ignore
}
```

and the following output in the cli
<img width="750" alt="image" src="https://user-images.githubusercontent.com/29753584/220577907-6e09f9b0-8b69-4d7d-b0cf-94862c7e1258.png">

Commit:
Updated type -> tpe for code 21

jkomyno committed:
test(migrate): test views warning code 21 for postgres

jkomyno committed:
test(migrate): fix typo

jkomyno committed:
test(migrate): add more tests for views in postgres

jkomyno committed:
test(migrate): minor refactoring after trying to centralize before/after Jest hooks

jkomyno committed:
test(migrate): add example of re-introspection of postgres warnings in views

Commit:
Added views re-introspection tests for codes 22-26
Updated test for code 21

Druue reviewed (commented):
Note: only code 24 and 26 show up during _introspection_
the rest (barring 22) only show up during _re-introspection_, this is currently as expected per internal discussion with @pimeys

Commit:
Updated test for views code 22

Druue commented:
22 is now tracked by test :)
![image](https://user-images.githubusercontent.com/29753584/221144617-d10d0f73-a267-46f4-a5b2-40b7a1c80271.png)

SevInf commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    })
+
+    beforeEach(async () => {
+      await setupPostgres(setupParams!).catch((e) => {
```
Are we sure we want to silence the error and proceed if setup fails? Chances that any of the tests will succeed in this case are pretty slim, imho.

Druue replied:
Hm, I'm less familiar as to what alternatives we have / what would be more idiomatic/helpful here

SevInf replied:
I'd just drop that `.catch`:
- if `setupPostgress` throws exception, we would still see it.
- if we could not setup postgress because of exception, running the rest of the tests is likely pointless - they won't find the database they'd expect.

Druue replied:
Would you recommend the same thing then for the `afterEach` `tearDown`?

SevInf commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    const setupParams = computeSetupParams(warningCode)
+
+    beforeAll(async () => {
+      await tearDownPostgres(setupParams!).catch((e) => {
```
I don't think we'll need this teardown here, as it should also happen in `afterEach` for all the tests.

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+  function computeSetupParams(warningCode: EngineArgs.ViewWarningCodes) {
+    const setupParams: SetupParams = {
+      connectionString,
+      // Note: at this location there is a setup.sql file
+      // which will be executed a SQL file so the database is not empty
```
```suggestion
      // which will be executed so the database is not empty
```

janpio replied:
```suggestion
      // Note: dirname points to a location with a setup.sql file
      // which will be executed to prepare the database with the correct tables, views etc.
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+  function computeSetupParams(warningCode: EngineArgs.ViewWarningCodes) {
+    const setupParams: SetupParams = {
+      connectionString,
+      // Note: at this location there is a setup.sql file
```
```suggestion
      // Note: dirname points to a location with a setup.sql file
```

SevInf commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+      process.env.TEST_POSTGRES_URI_MIGRATE = connectionString
+    })
+
+    afterEach(async () => {
```
Feel free to ignore the suggestion if you disagree, but i think this setup is repeated for all of the tests? In that case, it could also be split into helpers:

```ts
function setupPostgressForWarining(warningCode: EngineArgs.ViewWarningCodes) {
  const setupParams = computeSetupParams(warningCode)
  
  beforeEach(async () => {
      await setupPostgres(setupParams!).catch((e) => {
        console.error(e)
      })
      // Back to original env vars
      process.env = { ...originalEnv }
      // Update env var because it's the one that is used in the schemas tested
      process.env.TEST_POSTGRES_URI_MIGRATE = connectionString
    })

    afterEach(async () => {
      // Back to original env vars
      process.env = { ...originalEnv }
      await tearDownPostgres(setupParams!).catch((e) => {
        console.error(e)
      })
    })
}
```

Druue replied:
Yep, absolutely agree!

SevInf commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    })
+
+    beforeEach(async () => {
+      await setupPostgres(setupParams!).catch((e) => {
```
```suggestion
      await setupPostgres(setupParams).catch((e) => {
```

I don't think `setupParams` are nullable.

Druue replied:
As far as I can tell, they aren't, not sure what the reasoning was for this

SevInf commented on packages/migrate/src/types.ts:
```diff
@@ -322,6 +370,14 @@ export namespace EngineArgs {
     affected: AffectedModelOrCompositeTypeAndField[]
   }
 
+  export type ViewWarningCodes =
```
Unless I am missing something, this type is used only for test, so I would've moved it there.

Druue replied:
The reasoning for this was that the underlying types _aren't_ exported but I'm not sure if that's something that we can just update or if we should leave as is

SevInf replied:
Ah, ok then, I think in that case it's fine to keep it where it is.

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    return setupParams
+  }
+
+  describe('postgresql views 21', () => {
```
```suggestion
  describe('postgresql views 21 - unsupported types', () => {
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+  })
+
+  // TODO :: can't generate the correct code
+  describe('postgresql views 22', () => {
```
```suggestion
  describe('postgresql views 22 - @map', () => {
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    })
+  })
+
+  describe('postgresql views 23', () => {
```
```suggestion
  describe('postgresql views 23 - @@map', () => {
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    })
+  })
+
+  describe('postgresql views 24', () => {
```
```suggestion
  describe('postgresql views 24 - no unique identifier', () => {
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    })
+  })
+
+  describe('postgresql views 25', () => {
```
```suggestion
  describe('postgresql views 25 - @@id name', () => {
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    })
+  })
+
+  describe('postgresql views 26', () => {
```
```suggestion
  describe('postgresql views 26 - invalid field name', () => {
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
   })
 })
 
+describe('postgresql views re-introspection warnings', () => {
```
It's unexpected/surprising that these come before plain Introspection to me.

Druue replied:
This is what @jkomyno recommended when we paired on this 😅 Didn't think to ask about it at the time

janpio replied:
Maybe doesn't matter - it just took me a bit of scrolling to understand that the first big block is Re-Introspection (with existing schema file), and the second one is normal Introspection.

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    return setupParams
+  }
+
+  describe('postgresql views 21', () => {
```
```suggestion
  describe('postgresql views 21 - unsupported type (+ no unique identifier, because view)', () => {
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+  })
+
+  // TODO :: can't generate the correct code
+  describe('postgresql views 22', () => {
```
```suggestion
  describe('postgresql views 22 - no unique identifier', () => {
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    })
+  })
+
+  describe('postgresql views 23', () => {
```
```suggestion
  describe('postgresql views 23 - automated rename (+ no unique identifier, because view)', () => {
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    })
+  })
+
+  describe('postgresql views 24', () => {
```
What happens here that is specific to code 24?

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    })
+  })
+
+  describe('postgresql views 25', () => {
```
What happens here that is specific to code 25?

Druue replied:
Currently nothing, the original intent was to figure out what was happening with the cli warnings only emitting code 24 for introspection but given that it's intended for spam reasons (and in alignment with models), this could be removable now. Same goes for the introspecting 24 test

janpio replied:
I do not understand what "spam reasons" means. 24 and 25 have to do _something_ special that the others do not, or it does not make sense that they exist.

Druue replied:
Ah, that we don't emit too many warnings at once basically :)

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    })
+  })
+
+  describe('postgresql views 26', () => {
```
```suggestion
  describe('postgresql views 26 - invalid field name (+ no unique identifier, because view)', () => {
```

janpio commented on packages/migrate/src/__tests__/fixtures/introspection/postgresql/views-warning-26/setup.sql:
```diff
@@ -0,0 +1,5 @@
+-- TODO: remove tmp table once we no longer receive error P4001
```
I think this should already be resolved, right @pimeys?

Druue replied:
Yep, seems to be!

janpio commented on packages/migrate/src/__tests__/fixtures/re-introspection/postgresql/views-warning-22/setup.sql:
```diff
@@ -0,0 +1,6 @@
+-- TODO: remove tmp table once we no longer receive error P4001
```
see other comment

janpio commented on packages/migrate/src/__tests__/fixtures/re-introspection/postgresql/views-warning-23/setup.sql:
```diff
@@ -0,0 +1,6 @@
+-- TODO: remove tmp table once we no longer receive error P4001
```
see other comment

janpio commented on packages/migrate/src/__tests__/fixtures/re-introspection/postgresql/views-warning-25/setup.sql:
```diff
@@ -0,0 +1,6 @@
+-- TODO: remove tmp table once we no longer receive error P4001
```
see other comment

janpio commented on packages/migrate/src/__tests__/fixtures/re-introspection/postgresql/views-warning-26/setup.sql:
```diff
@@ -0,0 +1,6 @@
+-- TODO: remove tmp table once we no longer receive error P4001
```
see other comment

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    })
+  })
+
+  // TODO :: can't generate the correct code
```
??

Druue replied:
Ah, this can be removed now 😅

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,947 @@ describe('postgresql - missing database', () => {
+    })
+  })
+
+  // TODO :: can't generate the correct code
```
??

janpio commented on packages/migrate/src/__tests__/fixtures/introspection/postgresql/views-warning-22/setup.sql:
```diff
@@ -0,0 +1,5 @@
+-- TODO: remove tmp table once we no longer receive error P4001
```
see other comment

janpio commented on packages/migrate/src/__tests__/fixtures/introspection/postgresql/views-warning-23/setup.sql:
```diff
@@ -0,0 +1,5 @@
+-- TODO: remove tmp table once we no longer receive error P4001
```
see other comment

janpio commented on packages/migrate/src/__tests__/fixtures/introspection/postgresql/views-warning-25/setup.sql:
```diff
@@ -0,0 +1,5 @@
+-- TODO: remove tmp table once we no longer receive error P4001
```
see other comment

janpio reviewed (approved):
Looks good!

Left suggestion on naming the different tests to explain what the codes actually mean.

Commit:
update: review comments

Co-authored-by: Jan Piotrowski <piotrowski+github@gmail.com>

Commit:
Updated snapshots: remove tmp model

Commit:
Extract before / after setup/teardown to helper

Co-authored-by: Sergey Tatarintsev <tatarintsev@prisma.io>

Commit:
remove not-null enforcement on setupParams
- not nullable

Co-authored-by: Sergey Tatarintsev <tatarintsev@prisma.io>

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,669 @@ describe('postgresql - missing database', () => {
+    })
+  }
+
+  describe('postgresql views 21 - unsupported types', () => {
```
```suggestion
  describe('postgresql views 21 - unsupported types (both view and model [code 3])', () => {
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,669 @@ describe('postgresql - missing database', () => {
+    })
+  }
+
+  describe('postgresql views 21 - unsupported types', () => {
```
```suggestion
  describe('postgresql views 21 - unsupported types (both view and model [code 3])', () => {
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,669 @@ describe('postgresql - missing database', () => {
+
+    test('basic introspection', async () => {
+      ctx.fixture(`introspection/postgresql/views-warning-${warningCode}`)
+      const introspect = new DbPull()
+      const result = introspect.parse(['--print'])
+      await expect(result).resolves.toMatchInlineSnapshot(``)
+      expect(ctx.mocked['console.log'].mock.calls.join('\n')).toMatchInlineSnapshot(`
+        generator client {
+          provider        = "prisma-client-js"
+          previewFeatures = ["views"]
+        }
+
+        datasource db {
+          provider = "postgres"
+          url      = env("TEST_POSTGRES_URI_MIGRATE")
+        }
+
+        /// The underlying view does not contain a valid unique identifier and can therefore currently not be handled by the Prisma Client.
+        view A {
+          id Int?
+
+          @@ignore
+        }
+
+
+        // introspectionSchemaVersion: NonPrisma,
+      `)
+      expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+      expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`
+
+                                                                                                                                                        // *** WARNING ***
+                                                                                                                                                        // 
+                                                                                                                                                        // The following views were ignored as they do not have a valid unique identifier or id. This is currently not supported by the Prisma Client.
+                                                                                                                                                        // - View "A"
+                                                                                                                                                        // 
+                                                                                                                  `)
+      expect(ctx.mocked['process.stdout.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+      expect(ctx.mocked['process.stderr.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+    })
+  })
```
```suggestion
  // code 22 does not make sense in Introspection, as it requires a schema to pick up the prior @map
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,669 @@ describe('postgresql - missing database', () => {
+    })
+  })
+
+  describe('postgresql views 23 - view with @@map', () => {
```
```suggestion
  describe('postgresql views 23 - rename view with @@map', () => {
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,669 @@ describe('postgresql - missing database', () => {
+      expect(ctx.mocked['process.stderr.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+    })
+  })
+
```
```suggestion

// 23 is only re-introspection for keeping @@map from previous schema file

```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,669 @@ describe('postgresql - missing database', () => {
+      ctx.fixture(`introspection/postgresql/views-warning-${warningCode}`)
+      const introspect = new DbPull()
+      const result = introspect.parse(['--print'])
+      await expect(result).resolves.toMatchInlineSnapshot(``)
+      expect(ctx.mocked['console.log'].mock.calls.join('\n')).toMatchInlineSnapshot(`
+        generator client {
+          provider        = "prisma-client-js"
+          previewFeatures = ["views"]
+        }
+
+        datasource db {
+          provider = "postgres"
+          url      = env("TEST_POSTGRES_URI_MIGRATE")
+        }
+
+        /// This view has been renamed to 'Renamedif' during introspection, because the original name 'if' is reserved.
+        /// The underlying view does not contain a valid unique identifier and can therefore currently not be handled by the Prisma Client.
+        view Renamedif {
+          id Int?
+
+          @@map("if")
+          @@ignore
+        }
+
+
+        // introspectionSchemaVersion: NonPrisma,
+      `)
+      expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+      expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`
+
+                                                                                                                                                // *** WARNING ***
+                                                                                                                                                // 
+                                                                                                                                                // The following views were ignored as they do not have a valid unique identifier or id. This is currently not supported by the Prisma Client.
+                                                                                                                                                // - View "Renamedif"
+                                                                                                                                                // 
+                                                                                                            `)
+      expect(ctx.mocked['process.stdout.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+      expect(ctx.mocked['process.stderr.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+    })
+  })
```
Remove this temporarily while @pimeys actually adds a new code for this. 
```suggestion
```

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,669 @@ describe('postgresql - missing database', () => {
+    })
+  })
+
+  describe('postgresql views 24 - no unique identifier', () => {
```
TODO: Also test with multiple views to see that it does a list instead of 2 entries

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,669 @@ describe('postgresql - missing database', () => {
+    test('basic introspection', async () => {
+      ctx.fixture(`introspection/postgresql/views-warning-${warningCode}`)
+      const introspect = new DbPull()
+      const result = introspect.parse(['--print'])
+      await expect(result).resolves.toMatchInlineSnapshot(``)
+      expect(ctx.mocked['console.log'].mock.calls.join('\n')).toMatchInlineSnapshot(`
+        generator client {
+          provider        = "prisma-client-js"
+          previewFeatures = ["views"]
+        }
+
+        datasource db {
+          provider = "postgres"
+          url      = env("TEST_POSTGRES_URI_MIGRATE")
+        }
+
+        /// The underlying view does not contain a valid unique identifier and can therefore currently not be handled by the Prisma Client.
+        view B {
+          a Int?
+          b Int?
+
+          @@ignore
+        }
+
+
+        // introspectionSchemaVersion: NonPrisma,
+      `)
+      expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+      expect(ctx.mocked['console.error'].mock.calls.join('\n')).toMatchInlineSnapshot(`
+
+                                                                                                                                        // *** WARNING ***
+                                                                                                                                        // 
+                                                                                                                                        // The following views were ignored as they do not have a valid unique identifier or id. This is currently not supported by the Prisma Client.
+                                                                                                                                        // - View "B"
+                                                                                                                                        // 
+                                                                                                      `)
+      expect(ctx.mocked['process.stdout.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+      expect(ctx.mocked['process.stderr.write'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+    })
+  })
```
```suggestion

// 25 is re-introspection only to keep @@id names from previous schema 

```

Commit:
Update test names for 21 to be more specific
Removed un-necessary introspection tests
- added info comments for context

Co-authored-by: Jan Piotrowski <piotrowski+github@gmail.com>

Commit:
removed fixtures for removed tests

Commit:
updated fixtures for code 24
- added 'variant' to specifiy what to test
- new test for multiple code 24 warnings

Commit:
Removed catch for views (re-)introspection setup

Co-authored-by: Sergey Tatarintsev <tatarintsev@prisma.io>

janpio commented on packages/migrate/src/__tests__/DbPull.test.ts:
```diff
@@ -480,6 +481,601 @@ describe('postgresql - missing database', () => {
+    return setupParams
+  }
+
+  function setupPostgressForWarning(warningCode: EngineArgs.ViewWarningCodes) {
```
Too late, but `Postress` 😆
